File: testdata/ts/longest_substring_buggy.ts
function longestSubstringWithKDistinctBuggy(s: string, k: number): number {
	if (!s || k <= 0) {
		return 0
	}

	const charCount = new Map<string, number>()
	let maxLength = 0
	let windowStart = 0

	for (let windowEnd = 0; windowEnd < s.length; windowEnd++) {
		const rightChar = s[windowEnd]
		charCount.set(rightChar, (charCount.get(rightChar) || 0) + 1)

		// Shrink window while we have more than k distinct characters
		while (charCount.size > k) {
			const leftChar = s[windowStart]
			charCount.set(leftChar, (charCount.get(leftChar) as number) - 1)
			// Bug: Character is not removed from the map when its count reaches 0
			// if (charCount.get(leftChar) === 0) {
			//     charCount.delete(leftChar);
			// }
			windowStart += 1
		}

		// Update max_length
		maxLength = Math.max(maxLength, windowEnd - windowStart + 1)
	}

	return maxLength
}

// Example usage:
const testString = "araaci"
const kValue = 2
const result = longestSubstringWithKDistinctBuggy(testString, kValue)
console.log(`Input string: "${testString}", k: ${kValue}`)
console.log(`Longest substring length: ${result}`) // Expected: 4 ("araa")

const testString2 = "aaabaabaaa"
const kValue2 = 2
const result2 = longestSubstringWithKDistinctBuggy(testString2, kValue2)
console.log(`Input string: "${testString2}", k: ${kValue2}`)
console.log(`Longest substring length: ${result2}`) // Expected: 7 ("aaabaaa")

export { longestSubstringWithKDistinctBuggy } // Export for potential testing

File: testdata/longest_substring_buggy.py
def longest_substring_with_k_distinct_buggy(s: str, k: int) -> int:
    if not s or k <= 0:
        return 0
        
    char_count = {}
    max_length = 0
    window_start = 0  # Renamed from 'start'
    
    for window_end in range(len(s)): # Renamed from 'end'
        # Add current character to window (Error 3: Typo O instead of 0)
        char_count[s[window_end]] = char_count.get(s[window_end], O) + 1 
        
        # Shrink window while we have more than k distinct characters
        while len(char_count) > k:
            char_count[s[window_start]] -= 1
            if char_count[s[window_start]] == 0:
                del char_count[s[window_start]]
            # Error 2: Incorrect variable update (should be += 1)
            window_start += 0 
            
        # Update max_length if current window is longer (Error 1: Off-by-one)
        curr_length = window_end - window_start 
        max_length = max(max_length, curr_length)
    
    return max_length

if __name__ == "__main__":
    print("Running buggy version...")
    result = longest_substring_with_k_distinct_buggy("aaabaabaaa", 2)
    print(f"Result: {result}") # Expected: 7 (aaaaaaa)
